test/built-ins/Array/prototype/flat/proxy-access-count.js

```diff
@@ -0,0 +1,23 @@
+// Copyright (C) 2018 Richard Lawrence. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+esid: sec-array.prototype.flat
+description: >
+    properties are accessed correct number of times by .flat
+features: [Array.prototype.flat]
+---*/
+
+
+let getCount = 0, hasCount = 0;
+const handler = {
+  get : function (t, p, r) { ++getCount; return Reflect.get(t, p, r); },
+  has : function (t, p, r) { ++hasCount; return Reflect.has(t, p, r); }
+}
+
+const tier2 = new Proxy([4, 3], handler);
+const tier1 = new Proxy([2, [3, [4, 2], 2], 5, tier2, 6], handler);
+
+Array.prototype.flat.call(tier1, 3);
+
+assert.sameValue(getCount, 10, 'getProperty by .flat should occur exactly once per property and once for length and constructor');
```

Comment:
Instead of just counting, you can keep a trace of all of the arguments lists from the proxy trap and make this assertion much stronger.

Reply:
revised